### ISSUE TITLE
Selector matchers on some HTML fragments fail when nokogumbo is loaded

### ISSUE THREAD

**Issue by urbanautomaton** [teamcapybara/capybara#2137]
## Meta

Capybara Version: 3.12.0

## Expected Behavior

The following test should pass:

```ruby
require 'nokogumbo'

RSpec.describe 'asserting CSS in an HTML fragment' do
  include Capybara::DSL

  it 'finds a <td> element without a parent table' do
    expect(Capybara.string("<td></td>")).to have_css('td')
  end
end
```

## Actual Behavior

```
F

Failures:

  1) asserting CSS in an HTML fragment finds a <td> element without a parent
     Failure/Error: expect(Capybara.string("<td></td>")).to have_css('td')
       expected #has_css?("td") to return true, got false
     # ./spec/html_fragment_spec.rb:8:in `block (2 levels) in <top (required)>'

Finished in 0.02734 seconds (files took 0.58888 seconds to load)
1 example, 1 failure
```

## Steps to reproduce

I've set up a [repro over here](https://github.com/urbanautomaton/capybara_html_fragment_assertions).  To use:

    $ git clone https://github.com/urbanautomaton/capybara_html_fragment_assertions
    $ cd capybara_html_fragment_assertions
    $ bundle
    $ bundle exec rspec

## Cause

Since [version 3.11.0](https://github.com/teamcapybara/capybara/commit/7aa04bb5cf950e57bd4096ead5e0527ce185d0d2), capybara uses nokogumbo to parse HTML (via `Nokogiri::HTML5`) if the gem is loaded.

Nokogumbo appears to ignore nodes that don't have a required parent. For example, nokogumbo ignores `<td>` nodes unless they're nested within a `<table>` and `<tr>` or `<th>`, stripping them completely from the resulting AST. See the [repro readme](https://github.com/urbanautomaton/capybara_html_fragment_assertions#issue-description) for example AST output.

While nokogumbo supports Nokogiri's fragment-parsing interface, it has the same behaviour, and the `<td>` is still stripped.

## Possible fixes

I've reported this as an issue here as the change was [relatively recently introduced](https://github.com/teamcapybara/capybara/commit/7aa04bb5cf950e57bd4096ead5e0527ce185d0d2), and the behaviour isn't configurable. While there may be an issue with nokogumbo or the underlying gumbo parser, it's not obvious that their behaviour is incorrect. Meanwhile, the specific contexts capybara is used in mean it's relatively likely to be making assertions about HTML fragments (e.g. Rails template tests).

Two possible fixes I considered are:

* revert the change and ignore nokogumbo
* make the parser configurable

I didn't want to submit a PR without checking here first, though, as I don't know the reasons for using nokogumbo in the first place. I'd be very happy to work on one if we can agree an approach, though.

What do you think? I'd be very grateful for any and all opinions.

**Comment by twalpole** [comment]
Hmmm -- that's disappointing.  The reason the use of nokogumbo support was added was because gumbo does a much better job parsing HTML5 than libxml does - however when dealing with fragments it does appear it converts them into a document by wrapping with `<html><body>fragment</body></html>` which, in this case, then becomes illegal so the `<td>` is ignored.  I'm not sure how common of a use case it would be to be testing individual HTML elements that are technically required to have a  parent (ie - test the overall table template instead of the individual row template) but I'd be willing to accept a PR that adds a setting to disable the use of nokogumbo if it exists.

**Comment by twalpole** [comment]
By the way, thanks for the very clear issue report and reproduction 👍

**Comment by urbanautomaton** [comment]
Thanks for the response - I'll have a go at making the parser configurable. 👍

I know what you mean about the common-ness of the use case; our test suite maybe does have an unusual number of individually-tested partials in it. In the case I think I was looking at, we had a polymorphic table, so individually testing the partials generating the different row types made some sort of sense.

I think there's probably still value in having the parser configurable (not least cos having nokogumbo loaded may not be deliberate; in our app it's pulled in by the [sanitize gem](https://rubygems.org/gems/sanitize)), so am happy to do a PR.

**Comment by twalpole** [comment]
Fixed via f059e302f750aa10b545f3b041d93e087672cdaa

**Comment by urbanautomaton** [comment]
Oh! Thanks very much. Too slow. 😄

(Also, thank you very much for a wonderful library that makes our lives immensely easier. It's very much appreciated.)
